File: js/src/maplestory/api/common/dto/union/unionChampion.ts
export abstract class UnionChampionDto {
  public abstract date: Date | null;
  public abstract unionChampion: UnionChampionInfoDto[];
  public abstract championBadgeTotalInfo: UnionChampionBadgeInfoDto[];
}

export abstract class UnionChampionBadgeInfoDto {
  public abstract stat: string;
}

export abstract class UnionChampionInfoDto {
  public abstract championName: string;
  public abstract championSlot: number;
  public abstract championGrade: string;
  public abstract championClass: string;
  public abstract championBadgeInfo: UnionChampionBadgeInfoDto[];
}

File: csharp/MapleStory.OpenAPI/Src/Common/DTO/Union/UnionChampion.cs
namespace MapleStory.OpenAPI.Common.DTO
{
    public abstract class UnionChampionDTO<TUnionChampionInfo, TUnionChampionBadgeInfo>
    {
        public abstract DateTimeOffset? Date { get; set; }
        public abstract List<TUnionChampionInfo> UnionChampion { get; set; }
        public abstract List<TUnionChampionBadgeInfo> ChampionBadgeTotalInfo { get; set; }
    }

    public abstract class UnionChampionDTO : UnionChampionDTO<UnionChampionInfoDTO, UnionChampionBadgeInfoDTO>
    {

    }

    public abstract class UnionChampionBadgeInfoDTO
    {
        public abstract string Stat { get; set; }
    }

    public abstract class UnionChampionInfoDTO<TUnionChampionBadgeInfo>
    {
        public abstract string ChampionName { get; set; }
        public abstract int ChampionSlot { get; set; }
        public abstract string ChampionGrade { get; set; }
        public abstract string ChampionClass { get; set; }
        public abstract List<TUnionChampionBadgeInfo> ChampionBadgeInfo { get; set; }
    }

    public abstract class UnionChampionInfoDTO : UnionChampionInfoDTO<UnionChampionBadgeInfoDTO>
    {

    }
}
